<!DOCTYPE html>
<html lang="zh">
<head>
    <meta charset="UTF-8">
    <title>15-await的应用.html</title>
</head>
<body>
<script src="https://cdn.bootcdn.net/ajax/libs/jquery/3.6.0/jquery.min.js"></script>

<script>
    /*
        1. async修饰的函数
            函数的返回值为promise对象
            Promise实例的结果由async函数执行的返回值决定

        2. await表达式
            await右侧的表达式一般为Promise实例对象, 但也可以是其它的值
            (1).如果表达式是Promise实例对象, await后的返回值是promise成功的值
            (2).如果表达式是其它值, 直接将此值作为await的返回值

        3. 注意:
            await必须写在async函数中, 但async函数中可以没有await
            如果await的Promise实例对象失败了, 就会抛出异常, 需要通过try...catch来捕获处理
    */

    function sendAjax(url, data) {
        return new Promise((resolve, reject) => {
            //实例xhr
            const xhr = new XMLHttpRequest();
            //绑定监听
            xhr.onreadystatechange = () => {
                if (xhr.readyState === 4) {
                    if (xhr.status >= 200 && xhr.status < 300) {
                        resolve(xhr.response);
                    } else {
                        reject(`请求出了点问题`);
                    }
                }
            }
            //整理参数
            let str = '';
            for (let key in data) {
                str += `${key}=${data[key]}&`;
            }
            str = str.slice(0, -1);

            xhr.open('GET', url + '?' + str);
            xhr.responseType = 'json';
            xhr.send();
        });
    }

    (async () => {
        try {
            const result1 = await sendAjax('https://api.apiopen.top/getJoke', {page: 1});
            console.log('第1次请求成功了', result1);
            const result2 = await sendAjax('https://api.apiopen.top/getJoke', {page: 1});
            console.log('第2次请求成功了', result2);
            const result3 = await sendAjax('https://api.apiopen.top/getJoke', {page: 1});
            console.log('第3次请求成功了', result3);
        } catch (e) {
            console.log(e);
        }
    })();


</script>

</body>
</html>